### PR TITLE
Add doc page for Rewrite Prompts

### PR DESCRIPTION
**Pull request by TomeHirata**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/TomeHirata/mlflow/pull/18134?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18134/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18134/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/18134/merge
```

</p>
</details>

### Related Issues/PRs

N/A

### What changes are proposed in this pull request?

Add doc page for Rewrite Prompts

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [x] Examples
  - [ ] API references
  - [x] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+Your aim is to deliver a precise, correct answer that matches the expectation outlined.
+```
+
+## Components
+
+### 1. Predict Function
+
+Your `predict_fn` should:
+
+- Accept inputs as keyword arguments
```
Comment (harupy) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx:124]: Is this optional or required? `should` reads as optional or recommended. If it's required, let's use `must` to remove ambiguity.

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+    {"inputs": {"question": "What is the capital of Switzerland?"}, "outputs": "Bern"},
+    {"inputs": {"question": "What is the capital of Austria?"}, "outputs": "Vienna"},
+    {"inputs": {"question": "What is the capital of Belgium?"}, "outputs": "Brussels"},
+    {
+        "inputs": {"question": "What is the capital of Netherlands?"},
+        "outputs": "Amsterdam",
+    },
+    {"inputs": {"question": "What is the capital of Norway?"}, "outputs": "Oslo"},
+    {"inputs": {"question": "What is the capital of Sweden?"}, "outputs": "Stockholm"},
+]
```
Comment (harupy) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx:78]: can we create a `country -> capital` map and create a dataset from it?

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    def optimize(self, eval_fn, dataset, target_prompts, optimizer_lm_params):
+        # Your custom optimization logic
+        optimized_prompts = {}
+        for prompt_name, prompt_template in target_prompts.items():
+            # Implement your optimization algorithm
+            optimized_prompts[prompt_name] = your_optimization_logic(
+                prompt_template, dataset, optimizer_lm_params
+            )
+
+        return OptimizerAdapterOutput(optimized_prompts=optimized_prompts)
```
Comment (harupy) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx]: is this imported?

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+    ("Italy", "Rome"),
+    ("Spain", "Madrid"),
+    ("Portugal", "Lisbon"),
+    ("Switzerland", "Bern"),
+    ("Austria", "Vienna"),
+    ("Belgium", "Brussels"),
+    ("Netherlands", "Amsterdam"),
+    ("Norway", "Oslo"),
+    ("Sweden", "Stockholm"),
+]
```
Comment (harupy) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx]: can we use dict?

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+    {"question": "What is the capital of Germany?"},
+    {"question": "What is the capital of Japan?"},
+]
+
+for record in inputs:
+    predict_fn(**record)
+
+# Create your evaluation dataset
+dataset = create_dataset(
+    name="qa_evaluation_dataset",
```
Comment (harupy) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx]:
```suggestion
    name="qa_evaluation_dataset"
```

let's remove unnecessary `,`

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx [hunk 6]
```diff
@@ -0,0 +1,362 @@
+---
+sidebar_position: 6
+sidebar_label: Rewrite Prompts 🆕
+---
+
+import { APILink } from "@site/src/components/APILink";
+
+# Rewrite Prompts (Experimental)
```
Comment (harupy) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx]: is there a reason we use `rewrite` instead of `adapt`? `rewrite` might not convey the model-switching context.

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+- **Trace-Aware**: Leverages MLflow tracing to understand prompt usage patterns
+- **Flexible**: Works with any function that uses MLflow Prompt Registry
+
+:::
+
+## When to Use Prompt Rewriting
+
+Prompt rewriting is particularly useful when:
+
+- **Cost Optimization**: Downgrading to a cheaper model while maintaining quality
```
Comment (harupy) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx]:
```suggestion
- **Cost Optimization**: Downgrading to a more cost-effective model while preserving output quality
```

File: docs/sidebarsGenAI.ts [hunk 8]
```diff
@@ -380,6 +380,10 @@ const sidebarsGenAI: SidebarsConfig = {
           type: 'doc',
           id: 'prompt-registry/optimize-prompts',
         },
+        {
```
Comment (B-Step62) [inline docs/sidebarsGenAI.ts:383]: Can we group pages into some structure? Maybe we want to rename the top category to `Prompt Management` as well, since the functionalities it offers is far more than a registry.

Screenshot 2025-10-06 at 20 24 52

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+
+MLflow provides the <APILink fn="mlflow.genai.adapt_prompts" /> API to help you **maintain consistent outputs** when changing language models in your applications. This feature automatically rewrites your prompts to produce similar outputs with a new model as you were getting with your previous model. This API uses the [GEPA](https://arxiv.org/abs/2507.19457) optimization algorithm by default, but you can plug your custom algorithm as well.
+
+:::tip Key Benefits
+
+- **Model Migration**: Seamlessly switch between language models while maintaining output consistency
+- **Automatic Optimization**: Automatically rewrites prompts based on your existing data
+- **Trace-Aware**: Leverages MLflow tracing to understand prompt usage patterns
+- **Flexible**: Works with any function that uses MLflow Prompt Registry
+
```
Comment (B-Step62) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx:19]: Can we put some visuals in the doc? For example, comparison of before/after prompts, and a conceptual diagram that describes how this adaption works like [this](https://pr-18134--mlflow-docs-preview.netlify.app/docs/latest/genai/eval-monitor/scorers/llm-judge/agentic-overview/#how-it-works). It helps users understand the advanced concepts + encourage ppl to promote the feature in media.

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+prompt = mlflow.genai.register_prompt(
+    name="qa",
+    template="Answer the following question: {{question}}",
+)
+
+
+# Your current function
+def predict_fn(question: str) -> str:
+    completion = openai.OpenAI().chat.completions.create(
+        model="gpt-4o-mini",  # Target model
```
Comment (B-Step62) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx]:
```suggestion
        model="gpt-4o-mini",  # Target model to optimize the prompt for
```

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+```
+
+## Components
+
+### 1. Predict Function
+
+Your `predict_fn` must:
+
+- Accept inputs as keyword arguments
+- Use prompts from MLflow Prompt Registry via <APILink fn="mlflow.entities.PromptVersion.format" text="PromptVersion.format()" />
```
Comment (B-Step62) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx:125]: Out-of-scope: since we mandate users to include loaded prompt inside the `predict_fn`, is it possible to automatically detect which set of prompts to optimize? (unless specified via `target_prompt_uris`)

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx (hunk 9 above)
Comment (B-Step62) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx:19]:
```suggestion
- **Model Migration**: Seamlessly switch between language models while maintaining output consistency
- **No Ground Truth Needed**: MLflow automatically generates pseudo-labels from existing outputs, eliminating the need for manual annotation  
- **Automatic Optimization**: Automatically rewrites prompts based on your existing data
- **Trace-Aware**: Leverages MLflow tracing to understand prompt usage patterns
- **Flexible**: Works with any function that uses MLflow Prompt Registry
```

It is worth highlighting that this doesn't need human labeling. It seems the general impression ppl have with "optimization" is that these methods require manual labeling.

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx (hunk 11 above)
Comment (TomeHirata) [reply docs/docs/genai/prompt-registry/adapt-prompts.mdx:125]: I thought about it, but I decided not to do the inference since users can also have conditional prompt loading, but we need to know all initial prompts before optimization (like changing the prompt URL based on inputs).

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx (hunk 6 above)
Comment (TomeHirata) [reply docs/docs/genai/prompt-registry/adapt-prompts.mdx]: Good point, actually, I changed "adapt" to "rewrite" since I thought "Adapt prompt" is difficult to imagine the meaning. Do you think "Adapt prompt" is something many people can understand easily?

File: docs/sidebarsGenAI.ts (hunk 8 above)
Comment (TomeHirata) [reply docs/sidebarsGenAI.ts:383]: Let me rename the top category, but I'm not sure adding additional nesting is better here.

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx (hunk 9 above)
Comment (TomeHirata) [reply docs/docs/genai/prompt-registry/adapt-prompts.mdx:19]: That part is described in "Prepare Training Dataset from your existing application" section. I didn't include it in this top section since this API works with human labels and I didn't want to limit the usage for unlabeled cases. Maybe I'll mention it here with a bit more context instead of just saying "No Ground Truth Needed"

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx (hunk 6 above)
Comment (harupy) [reply docs/docs/genai/prompt-registry/adapt-prompts.mdx]: I'd prefer `Adapt`. `Rewrite` sounds a bit strange because whatever we do, we end up rewriting the prompt, right?

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx (hunk 9 above)
Comment (TomeHirata) [reply docs/docs/genai/prompt-registry/adapt-prompts.mdx:19]: Sounds good, I'm bad at visuals, but added SVG to explain the concept.

**Review by harupy — APPROVED** [review]
LGTM!

File: docs/docs/genai/prompt-registry/adapt-prompts.mdx [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+- **Trace-Aware**: Leverages MLflow tracing to understand prompt usage patterns
+- **Flexible**: Works with any function that uses MLflow Prompt Registry
+
+:::
+
+:::note[Version Requirements]
+The `adapt_prompts` API requires **MLflow >= 3.5.0**.
+:::
+
+![](/images/adapt_prompt_flow.svg)
```
Comment (harupy) [inline docs/docs/genai/prompt-registry/adapt-prompts.mdx:26]: Can we use mermaid instead? Never mind if its challenging

Comment (TomeHirata) [reply docs/docs/genai/prompt-registry/adapt-prompts.mdx:26]: I tried mermaid, but didn't get a good visual with it.
